filament/src/MorphTargets.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "details/MorphTargets.h"
+
+#include "private/filament/SibGenerator.h"
+
+#include "details/Engine.h"
+
+#include "FilamentAPI-impl.h"
+
+#include <math/mat4.h>
+
+#include <math/norm.h>
+
+namespace filament {
+
+using namespace backend;
+using namespace math;
+
+struct MorphTargets::BuilderDetails {
+    size_t mVertexCount = 0;
+    size_t mCount = 0;
+};
+
+using BuilderType = MorphTargets;
+BuilderType::Builder::Builder() noexcept = default;
+BuilderType::Builder::~Builder() noexcept = default;
+BuilderType::Builder::Builder(BuilderType::Builder const& rhs) noexcept = default;
+BuilderType::Builder::Builder(BuilderType::Builder&& rhs) noexcept = default;
+BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder const& rhs) noexcept = default;
+BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder&& rhs) noexcept = default;
+
+MorphTargets::Builder& MorphTargets::Builder::vertexCount(size_t vertexCount) noexcept {
+    mImpl->mVertexCount = vertexCount;
+    return *this;
+}
+
+MorphTargets::Builder& MorphTargets::Builder::count(size_t count) noexcept {
+    mImpl->mCount = count;
+    return *this;
+}
+
+MorphTargets* MorphTargets::Builder::build(Engine& engine) {
+    return upcast(engine).createMorphTargets(*this);
+}
+
+// ------------------------------------------------------------------------------------------------
+
+constexpr size_t MAX_TEXTURE_WIDTH = 4096;
```

Comment:
add a comment explaining where this is coming from. Also 4096 is not guaranteed to be supported everywhere. The GLES3 spec says:
```
GL_MAX_TEXTURE_SIZE
params returns one value. The value gives a rough estimate of the largest texture that the GL can handle. The value must be at least 2048. See glTexImage2D.
```
So unfortunately, this must be 2048.


On a side note, 2D arrays the numbers of layers is limited to 256 in the minspec:
```
GL_MAX_ARRAY_TEXTURE_LAYERS
params returns one value. The value indicates the maximum number of layers allowed in an array texture, and must be at least 256. See glTexImage2D.
```

Reply:
Thanks. I changed this value and add comments.